lib/netext/dialer.go

```diff
@@ -68,16 +67,23 @@ func (b BlackListedIPError) Error() string {
 
 // DialContext wraps the net.Dialer.DialContext and handles the k6 specifics
 func (d *Dialer) DialContext(ctx context.Context, proto, addr string) (net.Conn, error) {
-	delimiter := strings.LastIndex(addr, ":")
-	host := addr[:delimiter]
-
-	// lookup for domain defined in Hosts option before trying to resolve DNS.
-	ip, ok := d.Hosts[host]
-	if !ok {
-		var err error
-		ip, err = d.Resolver.FetchOne(host)
-		if err != nil {
-			return nil, err
+	host, port, err := net.SplitHostPort(addr)
+	if err != nil {
+		return nil, err
+	}
+
+	ip := net.ParseIP(host)
+	if ip == nil {
+		// It's not an IP address, so lookup the hostname in the Hosts
+		// option before trying to resolve DNS.
+		var ok bool
+		ip, ok = d.Hosts[host]
+		if !ok {
+			var dnsErr error
+			ip, dnsErr = d.Resolver.FetchOne(host)
+			if dnsErr != nil {
+				return nil, dnsErr
+			}
 		}
```

Comment:
This is probably good for a function and can be tested on it's own

Reply:
I agree, the `IP:port` splitting and DNS resolving logic can be a separate function(s), so you can testing it with unit tests with valid and invalid IPv4, IPv6, and domains. I also think we can also test the actual IPv6 connection logic as well, at least if we have IPv6 on our systems or CI. And we can reliably skip it if we don't.

We _should_ be able to use [`net.Listen`](https://golang.org/pkg/net/#Listen) with `tcp6` for `network` and port `0`. If it returns an error, then we `t.Skip()` the test. If not, we can spin up a test [`httptest.Server`](https://golang.org/pkg/net/http/httptest/#Server), or even httpmultibin there.

Reply:
> We should be able to use net.Listen with tcp6 for network and port 0. If it returns an error, then we t.Skip() the test. If not, we can spin up a test httptest.Server, or even httpmultibin there.

I suppose this could be done, I am not sure it will test everything nicely as we can probably test it on the loopback interface, but not between machines .... :shrug: But we can at least try that :D

Reply:
The loopback interface should still be employing most of the networking stack, I think - from the standpoint of Go and k6, everything should be pretty much the same, only the OS cares... That said, besides running the tests in the IPv6-enabled Docker container, like @imiric showed, might be the only way to run that test, for now, given that GitHub Actions doesn't support it: https://github.com/actions/virtual-environments/issues/668#issuecomment-624080758 :disappointed:

Reply:
Let me know if 437d91f1 is enough.

Ideally `resolveHost()` should work on a `Resolver` interface so that we can mock out the `net.LookupIP()` call. Right now this does an actual DNS lookup, which might fail if the DNS server doesn't support IPv6. Unlikely, but we should avoid environment issues like this, and there's no need to test the DNS. :)

Since we don't do much mocking elsewhere, it was easier to not do this, but let me know.

I'll look into the higher-level connection tests with `httpmultibin`.

Reply:
@imiric I vote for `t.Skip` on windows and unskipping it once on github actions for windows

Reply:
:-/ I'm leaning towards the interface approach, as it's not an issue with Windows, but with the environment. This test could fail just as easily on a dev machine using a non-IPv6 DNS, so removing the environment out of the equation is the safest approach for these unit tests. For integration tests there should be no mocking, but we can mark them as such and run them in an integration environment eventually.

Reply:
Take a look at d6c6ecc0.

I decided to extract more of `DialContext` into that function, so now it's awkwardly named `checkAndResolveAddress`, but I think having unit tests for this crucial part of k6 makes up for it.